api/src/main/java/com/epam/esm/hateoas/GiftCertificateDTOHateoas.java

```diff
@@ -0,0 +1,72 @@
+package com.epam.esm.hateoas;
+
+import com.epam.esm.controller.GiftCertificateController;
+import com.epam.esm.controller.TagController;
+import com.epam.esm.dto.GiftCertificateDTO;
+import com.epam.esm.dto.PaginationContainer;
+import com.epam.esm.dto.TagDTO;
+import com.epam.esm.dto.param.ParamContainer;
+import org.springframework.hateoas.CollectionModel;
+import org.springframework.hateoas.Link;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+import java.util.Set;
+
+import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo;
+import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.methodOn;
+
+/**
+ * Converter certificate according to HATEOAS rules
+ *
+ * @author Yauheni Tstiov
+ */
+@Component
+public class GiftCertificateDTOHateoas {
+
+    /**
+     * Converts the certificate according to the rules of HATEOAS.
+     *
+     * @param giftCertificateDTO contains a gift certificate for conversion.
+     */
+    public void build(GiftCertificateDTO giftCertificateDTO) {
+        Link self = linkTo(GiftCertificateController.class).slash(giftCertificateDTO.getId()).withSelfRel();
+        Link delete = linkTo(methodOn(GiftCertificateController.class)
```

Comment:
You shouldn't create a link for each HTTP method, it's redundant. You should create links that can be accessed from the current place, the client will decide by himself which HTTP method to use. According to your code, the will be dozens of similar links with different relations

Reply:
rework all the responses co contain selfLink + link to the internal list of objects, if needed
For example, getUser returns selfLink + link to `/orders` of the user

Reply:
Rework